File: 1038-Binary-Search-Tree-to-Greater-Sum-Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode bstToGst(TreeNode root) {
        greaterSum(root,0);
        return root;
    }

    public int greaterSum(TreeNode n, int sum){

        if (n.right!=null){
            sum=greaterSum(n.right,sum);
        }

        sum+=n.val;
        n.val=sum;

        if (n.left!=null){
            sum=greaterSum(n.left,sum);
        }

        return sum;
    }
}

File: LeetCode/1038-Binary-Search-Tree-to-Greater-Sum-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* bstToGst(TreeNode* root) {
        accumulate(root, 0);
        return root;
    }
    
    int accumulate(TreeNode* n, int sum){

        if (n->right != NULL) 
            sum = accumulate(n->right, sum);
        
        n->val += sum;
        sum = n->val ;

        if (n->left != NULL) 
            sum = accumulate(n->left, sum);

        return sum;
    }
};